crates/lunatic-networking-api/src/tls_tcp.rs

```diff
@@ -0,0 +1,878 @@
+//                  iterator with just one element and written to **id_ptr**.
+//
+// * 1 on error   - The error ID is written to **id_u64_ptr**
+//
+// Traps:
+// * If the tls listener ID doesn't exist.
+// * If any memory outside the guest heap space is referenced.
+fn tls_local_addr<T: NetworkingCtx + ErrorCtx>(
+    mut caller: Caller<T>,
+    tls_listener_id: u64,
+    id_u64_ptr: u32,
+) -> Result<u32, Trap> {
+    let tls_listener = caller
+        .data()
+        .tls_listener_resources()
+        .get(tls_listener_id)
+        .or_trap("lunatic::network::tls_local_addr: listener ID doesn't exist")?;
+    let (dns_iter_or_error_id, result) = match tls_listener.listener.local_addr() {
+        Ok(socket_addr) => {
+            let dns_iter_id = caller
+                .data_mut()
+                .dns_resources_mut()
+                .add(DnsIterator::new(vec![socket_addr].into_iter()));
+            (dns_iter_id, 0)
+        }
+        Err(error) => (caller.data_mut().error_resources_mut().add(error.into()), 1),
+    };
+
+    let memory = get_memory(&mut caller)?;
+    memory
+        .write(
+            &mut caller,
+            id_u64_ptr as usize,
+            &dns_iter_or_error_id.to_le_bytes(),
+        )
+        .or_trap("lunatic::network::tls_local_addr")?;
+
+    Ok(result)
+}
+
+// Creates a new TCP listener, which will be bound to the specified address. The returned listener
+// is ready for accepting connections.
+//
+// Binding with a port number of 0 will request that the OS assigns a port to this listener. The
+// port allocated can be queried via the `tls_local_addr` (TODO) method.
+//
+// Returns:
+// * 0 on success - The ID of the newly created TCP listener is written to **id_u64_ptr**
+// * 1 on error   - The error ID is written to **id_u64_ptr**
+//
+// Traps:
+// * If any memory outside the guest heap space is referenced.
+#[allow(clippy::too_many_arguments)]
+fn tls_bind<T: NetworkingCtx + ErrorCtx + Send>(
+    mut caller: Caller<T>,
+    addr_type: u32,
+    addr_u8_ptr: u32,
+    port: u32,
+    flow_info: u32,
+    scope_id: u32,
+    id_u64_ptr: u32,
+    certs_array_ptr: u32,
+    certs_array_len: u32,
+    keys_array_ptr: u32,
+    keys_array_len: u32,
+) -> Box<dyn Future<Output = Result<u32, Trap>> + Send + '_> {
+    Box::new(async move {
+        let memory = get_memory(&mut caller)?;
+        let certs = memory
+            .data(&caller)
+            .get(certs_array_ptr as usize..(certs_array_ptr + certs_array_len) as usize)
+            .or_trap("lunatic::networking::tls_accept")?
+            .to_vec();
+
+        let keys = memory
+            .data(&caller)
+            .get(keys_array_ptr as usize..(keys_array_ptr + keys_array_len) as usize)
+            .or_trap("lunatic::networking::tls_accept")?
+            .to_vec();
+        let keys = load_private_key(&keys).expect("should have unpacked the keys");
+        let certs = load_certs(&certs).expect("should have unpacked the certs");
```

Comment:
We shouldn't panic in host functions. You will need to return a trap in case of an error. In most cases you can use `or_trap` instead of `expect`.